src/components/street-segment.js

```diff
@@ -0,0 +1,424 @@
+        this.el.setAttribute(
+          `street-generated-pedestrians__${index}`,
+          `segmentWidth: ${this.data.width}; density: ${pedestrian.density}; length: ${this.data.length}; direction: ${this.data.direction};`
+        );
+      });
+    }
+  },
+  updateSurfaceFromType: function (typeObject) {
+    // update color, surface, level from segment type preset
+    this.el.setAttribute(
+      'street-segment',
+      `surface: ${typeObject.surface}; color: ${typeObject.color}; level: ${typeObject.level};`
+    ); // to do: this should be more elegant to check for undefined and set default values
+  },
+  updateGeneratedComponentsList: function () {
+    // get all components on entity with prefix 'street-generated'
+    let generatedComponentList = [];
+    const components = this.el.components;
+    for (const componentName in components) {
+      if (componentName.startsWith('street-generated')) {
+        generatedComponentList.push(componentName);
+      }
+    }
+    this.generatedComponents = generatedComponentList;
+  },
+  update: function (oldData) {
+    const data = this.data;
+    const dataDiff = AFRAME.utils.diff(oldData, data);
+    // if oldData is same as current data, then don't update
+    if (AFRAME.utils.deepEqual(oldData, data)) {
+      return;
+    }
+    // regenerate components if only type has changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('type')
+    ) {
+      let typeObject = this.types[this.data.type];
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      this.remove();
+      this.createGeneratedComponentsFromType(typeObject); // add components for this type
+      this.updateSurfaceFromType(typeObject); // update surface color, surface, level
+    }
+    // propagate change of direction to generated components is solo changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('direction')
+    ) {
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      for (const componentName of this.generatedComponents) {
+        this.el.setAttribute(componentName, 'direction', this.data.direction);
+      }
+    }
+    // propagate change of length to generated components is solo changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('length')
+    ) {
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      for (const componentName of this.generatedComponents) {
+        this.el.setAttribute(componentName, 'length', this.data.length);
+      }
+    }
+    this.clearMesh();
+    this.height = this.calculateHeight(data.level);
+    this.tempXPosition = this.el.getAttribute('position').x;
+    this.el.setAttribute('position', { x: this.tempXPosition, y: this.height });
```

Comment:
Please specify the z coordinates too or we'll have an issue in aframe 1.6.0.

Reply:
fcb5dcb438e67ad4cff2f86d8c567f23a00905f0